lib/wibox/init.lua

```diff
@@ -61,6 +63,72 @@ function wibox:find_widgets(x, y)
     return self._drawable:find_widgets(x, y)
 end
 
+-- This function is called on every redraw after the background has been drawn,
+-- but before the actual widget was drawn. We use this to draw an anti-aliased
+-- clipping shaped (the edge between the border and the background is
+-- anti-aliased)
+local function draw_hook(self, cr)
+    local shape = self._shape
+    if not shape then return end
+
+    local geo = self:geometry()
+    cr:set_source(gcolor(self.border_color))
+
+    -- We want to invert the shape (draw to anything that is not inside of it).
+    -- To fake this, we use EVEN_ODD and a rectangle containing the full
+    -- surface. Pixels in the border will be inside the big rectangle, so they
+    -- are contained once in the path. Pixels in the "content area" are
+    -- additionally contained in the shape, so twice.
+    cr:set_fill_rule(cairo.FillRule.EVEN_ODD)
+    cr:rectangle(0, 0, geo.width, geo.height)
+    shape(cr, geo.width, geo.height)
+
+    cr:fill()
+end
+
+function wibox:_apply_shape()
+    local shape = self._shape
+
+    if not shape then
+        self.shape_bounding = nil
+        self.shape_clip = nil
+        return
+    end
+
+    local geo = self:geometry()
+    local bw = self.border_width
+
+    -- First handle the bounding shape (things including the border)
+    local img = cairo.ImageSurface(cairo.Format.A1, geo.width + 2*bw, geo.height + 2*bw)
+    local cr = cairo.Context(img)
+
+    shape(cr, geo.width + 2*bw, geo.height + 2*bw)
```

Comment:
Ok, here is the patch and a screenshot:
```diff
diff --git a/awesomerc.lua b/awesomerc.lua
index 87d9da7..06ecdb4 100755
--- a/awesomerc.lua
+++ b/awesomerc.lua
@@ -210,6 +210,7 @@ awful.screen.connect_for_each_screen(function(s)
 
     -- @DOC_WIBAR@
     -- Create the wibox
+    --[[
     s.mywibox = awful.wibar({ position = "top", screen = s })
 
     -- @DOC_SETUP_WIDGETS@
@@ -231,6 +232,18 @@ awful.screen.connect_for_each_screen(function(s)
             s.mylayoutbox,
         },
     }
+    --]]
+    s.mywibox = wibox {
+        x = 100,
+        y = 100,
+        width = 300,
+        height = 300,
+        bg = "#ff0000",
+        border_color = "#00ff00",
+        border_width = 30,
+        shape = gears.shape.circle,
+    }
+    s.mywibox.visible = true
 end)
 -- }}}
 
diff --git a/lib/wibox/init.lua b/lib/wibox/init.lua
index 622734a..24af541 100644
--- a/lib/wibox/init.lua
+++ b/lib/wibox/init.lua
@@ -74,13 +74,28 @@ local function draw_hook(self, cr)
     local geo = self:geometry()
     local bw = self.border_width
 
+    -- Draw the border shape
     cr:translate(-bw, -bw)
     shape(cr, geo.width + 2*bw, geo.height + 2*bw)
+    cr:translate(bw, bw)
 
     -- 2*, because half of them is outside of the visible area
     cr:set_line_width(2*bw)
     cr:set_source(gcolor(self.border_color))
-    cr:stroke()
+    cr:stroke_preserve()
+
+    -- Now do magic to simulate anti-aliasing for the bounding shape. We
+    -- "invert" the shape via EVEN_ODD and adding a rectangle around everything.
+    -- Pixels inside of the shape will be contained twice in the path
+    -- (rectangle+shape), so an even number of times. Pixels only inside the
+    -- rectangle get an odd number.
+    -- TODO: Only if compositing manager is running
+    cr:rectangle(0, 0, geo.width, geo.height)
+    cr:set_fill_rule(cairo.FillRule.EVEN_ODD)
+    cr:set_operator(cairo.Operator.CLEAR)
+    cr:fill()
+
+    cr:new_path()
 end
 
 function wibox:_apply_shape()
```
![foo](https://cloud.githubusercontent.com/assets/89482/19835542/71725dee-9e8a-11e6-8931-da40c995364c.png)
If you zoom, you can see that this does have some effect, but I wouldn't say that the result really is good:
![foo-zoomed](https://cloud.githubusercontent.com/assets/89482/19835543/8add8182-9e8a-11e6-8cea-0f104f0794f4.png)
I don't like the hack from your version were you try magic with the line width, as this causes the artifacts that are visible in the screenshot I posted there.

Should we really try to make X11 do something that it is not meant to do?